### ISSUE TITLE
Can be download new beta version and other preparation for the new beta be automatized in the program?

### ISSUE THREAD

**Issue by Cyberyoda1411** [SubtitleEdit/subtitleedit#7390]
**NOTE. I am totally satisfied with my CMD program, I finish everything in one click. This is my attempt to help others.** 

Everyone who is smart, uses the latest beta version, with the newest features. It has no installation, it is in one folder. The steps you must do EVERY time when a new beta version is released:
1. You must establish if it is the newest version (you must do it by yourself, sorry)
2. Download it to your desktop (the best place).
3. You make a folder and you copy your zip beta file.
4. From your old beta folder, if you are smart, you copy 2 folders and 1 file in your new beta folder : ffmpeg and Whisper folders (or you will have to download it with every new beta version from scratch) and settings.xml (to save your settings, or you will spend 5 minutes to set everything you want, every time)
5. You unpack your new zip beta and make the shortcut of the file SubtitleEdit.exe
6. You delete the old beta version to save your space on the C partition.

All this can be automatized. Except the points 1 and 5. Sorry, I don't know how to make a shortcut from the command line. It simply doesn't work, I tried the mklink command, but without success.

CMD (or bat) file is:
-------------------------------------------------------------------------------------------------------
@echo off
REM this is my path to my desktop, for you, it will be something different than Dragan (it is your user name)
cd C:\Users\Dragan\Desktop

REM creating folder SEbetaT (T - temporary) and with wget you download the newest zip file from github
md SEbetaT
cd SEbetaT
wget --no-check-certificate -O SubtitleEditBeta.zip  https://github.com/SubtitleEdit/subtitleedit/releases/download/4.0.0/SubtitleEditBeta.zip

REM moving the downloaded beta zip file into the new temp folder
move C:\Users\Dragan\Desktop\SubtitleEditBeta.zip C:\Users\Dragan\Desktop\SEbetaT
cd C:\Users\Dragan\Desktop\SEbetaT

REM unpacking the newest beta zip file
tar -xf SubtitleEditBeta.zip

REM copying 2 folders (ffmpeg and Whisper) + 1 file (settings.xml) from the old beta which is stored in SEbeta
REM into SEbetaT (temporary folder)
md C:\Users\Dragan\Desktop\SEbetaT\ffmpeg
xcopy "C:\Users\Dragan\Desktop\SEbeta\ffmpeg" "C:\Users\Dragan\Desktop\SEbetaT\ffmpeg" /E /I /Y

md C:\Users\Dragan\Desktop\SEbetaT\Whisper
xcopy "C:\Users\Dragan\Desktop\SEbeta\Whisper" "C:\Users\Dragan\Desktop\SEbetaT\Whisper" /E /I /Y

xcopy "C:\Users\Dragan\Desktop\SEbeta\Settings.xml" "C:\Users\Dragan\Desktop\SEbetaT"

REM deleting old beta folder SEbeta, deleting the old link to SubtitleEdit.exe, and renaming SEbetaT into SEbeta 
cd C:\Users\Dragan\Desktop
rd /s /q foldername SEbeta
del SubtitleEdit.exe.lnk
rename SEbetaT SEbeta
------------------------------------------------------------------------------------------------------
Of course, Nikolaj can make this much more sophisticated in C#, with the input parameter. The program will ask for the User name in the path: C:\Users\User name\Desktop. This way you will enter your user name and the program will be universal. Even this CMD program can work with parameters, but I didn't need to make it that way for my use, my user name is a constant. 

I AM TOTALLY SATISFIED WITH MY CMD program, it makes my life much easier when the new beta is there, I finish everything in one click (I just make a shortcut to exe file and that is all in two steps), but I'd like **to help others to have the same benefit.** As I said, every real user will use the beta, because every new beta is better than the official version with the standard installation.

Even if Nikolaj consider this as not important, this CMD program can be shared and likned to Subtitle Edit, and people just need to change the user name in the program, copy this into some txt file, make a name for this, and rename extension in BAT or CMD.

**Comment by rRobis** [comment]
Hi!

I got interested in solving this to be automated 😅

This downloads latest beta.
Backups settings, libmpv-2.dll, Whisper and ffmpeg folder to desktop folder
Then it removes/makes a dir in C: disk SubtitleEditBeta and extracts the zip using tar.exe.

Then it copies the backups to this folder

And then it makes desktop shortcut of SubtitleEdit.exe from beta folder.

I'm not very experienced in this, but I tested, seems to be working fine. 


InstallSubtitleEditBeta.bat
```
@echo off

set "BetaFolder=C:\SubtitleEditBeta\"

:: Set current dir to Desktop
cd %UserProfile%\Desktop\

:: Download latest version of SubtitleEditBeta.ZIP
echo Downloading latest SubtitleEditBeta

curl -O -L "https://github.com/SubtitleEdit/subtitleedit/releases/latest/download/SubtitleEditBeta.zip"
cls
:: Create Backup Folder
mkdir %UserProfile%\Desktop\SettingsBackup
cls
:: Backup ffmpeg, Whisper, libmpv-2.dll and Settings.xml

:: Uncomment to use Stable Version settings / ffmpeg, mvp-2 library and Whisper and comment block below
set "StableFolder=%UserProfile%\AppData\Roaming\Subtitle Edit\"
::Xcopy "%StableFolder%\ffmpeg" "%UserProfile%\Desktop\SettingsBackup\ffmpeg" /I /E /Y
::Xcopy "%StableFolder%\Whisper" "%UserProfile%\Desktop\SettingsBackup\Whisper" /I /E /Y
::Xcopy "%StableFolder%\libmpv-2.dll" "%UserProfile%\Desktop\SettingsBackup" /Y
::Xcopy "%StableFolder%\Settings.xml" "%UserProfile%\Desktop\SettingsBackup" /Y

:: Comment this if you want to use Stable version backup
set "BetaFolder=C:\SubtitleEditBeta\"
Xcopy "%BetaFolder%\ffmpeg" "%UserProfile%\Desktop\SettingsBackup\ffmpeg" /I /E /Y
Xcopy "%BetaFolder%\Whisper" "%UserProfile%\Desktop\SettingsBackup\Whisper" /I /E /Y
Xcopy "%BetaFolder%\libmpv-2.dll" "%UserProfile%\Desktop\SettingsBackup" /Y
Xcopy "%BetaFolder%\Settings.xml" "%UserProfile%\Desktop\SettingsBackup" /Y

if exist "%BetaFolder%" (
rmdir /s /q %BetaFolder%
)

mkdir C:\SubtitleEditBeta\

tar -xf .\SubtitleEditBeta.zip -C "C:\SubtitleEditBeta"

:: Copy backup to beta folder
Xcopy "%UserProfile%\Desktop\SettingsBackup\ffmpeg" "%BetaFolder%\ffmpeg" /I /E /Y
Xcopy "%UserProfile%\Desktop\SettingsBackup\Whisper" "%BetaFolder%\Whisper" /I /E /Y
Xcopy "%UserProfile%\Desktop\SettingsBackup\libmpv-2.dll" "%BetaFolder%" /Y
Xcopy "%UserProfile%\Desktop\SettingsBackup\Settings.xml" "%BetaFolder%" /Y

:: Shortcut directories
set "TargetFile=C:\SubtitleEditBeta\SubtitleEdit.exe"
set "ShortcutName=SubtitleEditBeta.lnk"

:: Create the shortcut
echo Set WshShell = CreateObject("WScript.Shell")> CreateShortcut.vbs
echo Set Shortcut = WshShell.CreateShortcut(WshShell.SpecialFolders("Desktop") ^& "\" ^& "%ShortcutName%")>> CreateShortcut.vbs
echo Shortcut.TargetPath = "%TargetFile%">> CreateShortcut.vbs
echo Shortcut.Save>> CreateShortcut.vbs

:: Run the VBScript to create the shortcut
cscript CreateShortcut.vbs

:: Clean up the VBScript file and ZIP file
del CreateShortcut.vbs
del SubtitleEditBeta.zip
```

**Comment by rRobis** [comment]
Here is also a PowerShell function of same bat file, that you could put in your Powershell profile and just run it from terminal

You can rename iSEB as you want.

```
function iSEB {
    $BetaFolder = "C:\SubtitleEditBeta\"
    
    # Set current dir to Desktop
    Set-Location $env:UserProfile\Desktop

    # Download latest version of SubtitleEditBeta.ZIP
    Write-Host "Downloading latest SubtitleEditBeta"
    Invoke-WebRequest -Uri "https://github.com/SubtitleEdit/subtitleedit/releases/latest/download/SubtitleEditBeta.zip" -OutFile "SubtitleEditBeta.zip"

    Clear-Host

    # Create Backup Folder
    New-Item -ItemType Directory -Path "$env:UserProfile\Desktop\SettingsBackup" -Force

    Clear-Host

    # Backup ffmpeg, Whisper, libmpv-2.dll, and Settings.xml

    # Uncomment to use Stable Version settings / ffmpeg, mvp-2 library, and Whisper and comment block below
    $StableFolder = "$env:UserProfile\AppData\Roaming\Subtitle Edit\"
    #Copy-Item -Path "$StableFolder\ffmpeg" -Destination "$env:UserProfile\Desktop\SettingsBackup\ffmpeg" -Recurse -Force
    #Copy-Item -Path "$StableFolder\Whisper" -Destination "$env:UserProfile\Desktop\SettingsBackup\Whisper" -Recurse -Force
    #Copy-Item -Path "$StableFolder\libmpv-2.dll" -Destination "$env:UserProfile\Desktop\SettingsBackup" -Force
    #Copy-Item -Path "$StableFolder\Settings.xml" -Destination "$env:UserProfile\Desktop\SettingsBackup" -Force

    # Comment this if you want to use Stable version backup
    $BetaFolder = "C:\SubtitleEditBeta\"
    Copy-Item -Path "$BetaFolder\ffmpeg" -Destination "$env:UserProfile\Desktop\SettingsBackup\ffmpeg" -Recurse -Force
    Copy-Item -Path "$BetaFolder\Whisper" -Destination "$env:UserProfile\Desktop\SettingsBackup\Whisper" -Recurse -Force
    Copy-Item -Path "$BetaFolder\libmpv-2.dll" -Destination "$env:UserProfile\Desktop\SettingsBackup" -Force
    Copy-Item -Path "$BetaFolder\Settings.xml" -Destination "$env:UserProfile\Desktop\SettingsBackup" -Force

    if (Test-Path -Path $BetaFolder -PathType Container) {
        Remove-Item -Path $BetaFolder -Recurse -Force
    }

    New-Item -ItemType Directory -Path "C:\SubtitleEditBeta\"

    Expand-Archive -Path ".\SubtitleEditBeta.zip" -DestinationPath "C:\SubtitleEditBeta"

    # Copy backup to beta folder
    Copy-Item -Path "$env:UserProfile\Desktop\SettingsBackup\ffmpeg" -Destination "$BetaFolder\ffmpeg" -Recurse -Force
    Copy-Item -Path "$env:UserProfile\Desktop\SettingsBackup\Whisper" -Destination "$BetaFolder\Whisper" -Recurse -Force
    Copy-Item -Path "$env:UserProfile\Desktop\SettingsBackup\libmpv-2.dll" -Destination $BetaFolder -Force
    Copy-Item -Path "$env:UserProfile\Desktop\SettingsBackup\Settings.xml" -Destination $BetaFolder -Force

    # Create the shortcut
    $ShortcutFile = Join-Path $env:UserProfile\Desktop "SubtitleEditBeta.lnk"
    $ShortcutTarget = Join-Path $BetaFolder "SubtitleEdit.exe"
    $WScriptShell = New-Object -ComObject WScript.Shell
    $Shortcut = $WScriptShell.CreateShortcut($ShortcutFile)
    $Shortcut.TargetPath = $ShortcutTarget
    $Shortcut.Save()

# Clean up the ZIP file
	Remove-Item -Path ".\SubtitleEditBeta.zip" -Force
	Clear-Host
	Write-Host "All Done"
}
```

**Comment by rRobis** [comment]
Made some edits to previous 2 scripts.

**Comment by Cyberyoda1411** [comment]
Wow! I didn't know that you could use If and other nice things in the cmd or bat file. And I was right that it is not possible to make a shortcut just by mklink command. It is a bit more complex. 

This whole idea occurred to me spontaneously. I solved one line at a time and made automatization of the whole process, except for making a shortcut. I made a lot of mistakes, but I solved one by one problem. And I was right that people will be interested in doing the same thing because there are too many things to download, copy, make new folders, and unzip,..., it is too exhausting if you have (sometimes) 3 beta versions a day. Even if any of this doesn't become part of the program, it is great we all share solutions to make our lives easier. Maybe it can be part of the GitHub programs that come along with SE. 

And it is not about 'testing beta', because, in the beta versions, there is a lot of new functionality, it is stupid to use the official versions only, beta versions are much more advanced in any way. They are not just fixing bugs, they have some crucial new things.

**Comment by Cyberyoda1411** [comment]
Robis, I have two questions. 
1. Where did you find your path: 
https://github.com/SubtitleEdit/subtitleedit/releases/latest/download/SubtitleEditBeta.zip

which WORKS FINE. But on the page it is: 
https://github.com/SubtitleEdit/subtitleedit/releases/download/4.0.1/SubtitleEditBeta.zip

How did you find this first address, which gives you the latest beta version? It is from some other location? Because the GitHub releases list is only this one (which I managed to find), with the 4.0.1 folder and I must edit it every time the new version is released, while your URL is totally independent. What is your magic?

2. Why your curl command doesn't work on my computer? To be honest, the curl command doesn't work on my computer at all. Your command is:

curl -O -L "https://github.com/SubtitleEdit/subtitleedit/releases/latest/download/SubtitleEditBeta.zip"

I get nothing (I use wget, it works for me). 

When I try to find out if I have curl on my system, I get version 8.0.1 (it comes with Windows). It simply won't work. And when I try to install the newest version 8.3.0_1, the system won't let me delete curl.exe from Windows\system32. I set the path to curl.exe, but it shows the old 8.0.1 version. 

Any suggestions? I tried to rename curl.exe to curl_old.exe and copy the newest curl.exe, but the system doesn't let me do it. And some sources say that I shouldn't play with these things, because I can mess something up in Windows.

**Comment by rRobis** [comment]
1. I found out about it in StackOverflow, but now i found it also on github docs. You can link any latest release of any repository and point to a file and it will give you it.

https://docs.github.com/en/repositories/releasing-projects-on-github/linking-to-releases

2. About curl, I have no idea why. But if wget works, there's no reason to not replace it. I wanted to use wget on my pc, but found out that it's not installed, so i used curl, because it was.

**Comment by Cyberyoda1411** [comment]
I reinstalled my windows, and my cmd file doesn't work. OK, I needed to add 7z.exe (to unzip the zip file) and wget (to get the file from SE page) into PATH. But, xcopy command doesn't work. It should be

xcopy "PATH TO THE SOURCE FOLDERS" "PATH TO THE DESTINATION"  /E /I /Y

But it doesn't copy ANYTHING. Can anyone help me? Why it doesn't work? It doesn't copy a single directory, I didn't change anything, it worked on my last Windows installation.
